Error during the `asan` CI job: "LLVM ERROR: inconsistency in registered CommandLine options"

This issue is to track the progress on fixing the "LLVM ERROR: inconsistency in registered CommandLine options" error seen in the `asan` CI job.

```
    LINK tmp/test-asan/asan/usr/lib/libjulia-codegen-debug.so.1
    LINK tmp/test-asan/asan/usr/lib/libjulia-codegen-debug.so
make[1]: Leaving directory '/cache/build/amdci8-0/julialang/julia-master/tmp/test-asan/asan/src'
make[1]: Entering directory '/cache/build/amdci8-0/julialang/julia-master/tmp/test-asan/asan'
    JULIA tmp/test-asan/asan/usr/lib/julia/corecompiler.ji
: CommandLine Error: Option 'help-list' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
Aborted
make[1]: *** [/cache/build/amdci8-0/julialang/julia-master/sysimage.mk:61: /cache/build/amdci8-0/julialang/julia-master/tmp/test-asan/asan/usr/lib/julia/corecompiler.ji] Error 134
make[1]: Leaving directory '/cache/build/amdci8-0/julialang/julia-master/tmp/test-asan/asan'
make: *** [/cache/build/amdci8-0/julialang/julia-master/Makefile:82: julia-sysimg-ji] Error 2
make: Leaving directory '/cache/build/amdci8-0/julialang/julia-master/tmp/test-asan/asan'
🚨 Error: The command exited with status 2
```

Example log: https://buildkite.com/julialang/julia-master/builds/4455#3d8f2204-c0ae-43b8-a232-cfeb2a5be0c6

Introduced in https://github.com/JuliaLang/julia/pull/41936

--- a/.buildkite/pipelines/main/misc/sanitizers.yml
+++ b/.buildkite/pipelines/main/misc/sanitizers.yml
@@ -21,7 +21,6 @@ steps:
     timeout_in_minutes: 120
     if: | # We only run the `asan` job on Julia 1.8 and later.
       (pipeline.slug != "julia-release-1-dot-6") && (pipeline.slug != "julia-release-1-dot-7")
-    soft_fail: true # TODO: delete this line (and thus disallow failures) once JuliaLang/julia#42540 is fixed
     commands: |
       echo "--- Build julia-debug with ASAN"
       contrib/asan/build.sh ./tmp/test-asan -j$${JULIA_CPU_THREADS:?} debug

--- a/Make.inc
+++ b/Make.inc
@@ -483,9 +483,6 @@ endif
 endif
 
 ifeq ($(USEGCC),1)
-ifeq ($(SANITIZE),1)
-$(error Sanitizers are only supported with clang. Try setting SANITIZE=0)
-endif
 CC := $(CROSS_COMPILE)gcc
 CXX := $(CROSS_COMPILE)g++
 JCFLAGS := -std=gnu99 -pipe $(fPIC) -fno-strict-aliasing -D_FILE_OFFSET_BITS=64
@@ -518,6 +515,8 @@ export MACOSX_DEPLOYMENT_TARGET=$(MACOSX_VERSION_MIN)
 JCPPFLAGS += -D_LARGEFILE_SOURCE -D_DARWIN_USE_64_BIT_INODE=1
 endif
 endif
+
+JLDFLAGS :=
 
 ifeq ($(USECCACHE), 1)
 # Expand CC, CXX and FC here already because we want the original definition and not the ccache version.
@@ -1237,14 +1236,10 @@ IFUNC_DETECT_SRC := 'void (*f0(void))(void) { return (void(*)(void))0L; }; void 
 ifeq (supported, $(shell echo $(IFUNC_DETECT_SRC) | $(CC) -Werror -x c - -S -o /dev/null > /dev/null 2>&1 && echo supported))
 JCPPFLAGS += -DJULIA_HAS_IFUNC_SUPPORT=1
 endif
-JLDFLAGS := -Wl,-Bdynamic
-ifneq ($(SANITIZE),1)
-ifneq ($(SANITIZE_MEMORY),1)
-ifneq ($(LLVM_SANITIZE),1)
+JLDFLAGS += -Wl,-Bdynamic
 OSLIBS += -Wl,--version-script=$(JULIAHOME)/src/julia.expmap
+ifneq ($(SANITIZE),1)
 JLDFLAGS += -Wl,-no-undefined
-endif
-endif
 endif
 ifeq (-Bsymbolic-functions, $(shell $(LD) --help | grep -o -e "-Bsymbolic-functions"))
 JLIBLDFLAGS := -Wl,-Bsymbolic-functions
@@ -1258,7 +1253,7 @@ JLIBLDFLAGS :=
 endif
 
 ifeq ($(OS), FreeBSD)
-JLDFLAGS := -Wl,-Bdynamic
+JLDFLAGS += -Wl,-Bdynamic
 OSLIBS += -lelf -lkvm -lrt -lpthread -latomic
 
 # Tweak order of libgcc_s in DT_NEEDED,
@@ -1276,7 +1271,6 @@ SHLIB_EXT := dylib
 OSLIBS += -framework CoreFoundation
 WHOLE_ARCHIVE := -Xlinker -all_load
 NO_WHOLE_ARCHIVE :=
-JLDFLAGS :=
 HAVE_SSP := 1
 JLIBLDFLAGS := -Wl,-compatibility_version,$(SOMAJOR) -Wl,-current_version,$(JULIA_MAJOR_VERSION).$(JULIA_MINOR_VERSION).$(JULIA_PATCH_VERSION)
 endif
@@ -1285,7 +1279,7 @@ ifeq ($(OS), WINNT)
 HAVE_SSP := 1
 OSLIBS += -Wl,--export-all-symbols -Wl,--version-script=$(JULIAHOME)/src/julia.expmap \
 	$(NO_WHOLE_ARCHIVE) -lpsapi -lkernel32 -lws2_32 -liphlpapi -lwinmm -ldbghelp -luserenv -lsecur32 -latomic
-JLDFLAGS := -Wl,--stack,8388608
+JLDFLAGS += -Wl,--stack,8388608
 ifeq ($(ARCH),i686)
 JLDFLAGS += -Wl,--large-address-aware
 endif

--- a/src/julia.expmap
+++ b/src/julia.expmap
@@ -1,7 +1,5 @@
 {
   global:
-    __asan*;
-    __tsan*;
     pthread*;
     __stack_chk_guard;
     asprintf;

